File: ggcommons/heartbeat/heartbeat_monitor.py
import shutil
import psutil
import os
import platform
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ggcommons.config.manager.config_manager import ConfigManager


class HeartbeatMonitor:
    def __init__(self, config_service: "ConfigManager"):
        self._config_service = config_service
        self._config = config_service.get_heartbeat_config()
        self._pid = None
        self._proc_info = None
        self._platform = platform.system()
        self.pid, self.proc_info = HeartbeatMonitor.build_proc_info()

    def get_stats(self):
        data = {}
        cpu_data = self.cpu_usage()
        memory_data = self.memory_usage()
        disk_data = self.disk_usage()
        thread_data = self.thread_count()
        files_data = self.open_files()
        fds = self.file_descriptors()
        # Check for conflicting configurations
        if cpu_data is not None:
            data["cpu"] = cpu_data
        if memory_data is not None:
            data["memory"] = memory_data
        if disk_data is not None:
            data["disk"] = disk_data
        if thread_data is not None:
            data["threads"] = thread_data
        if files_data is not None:
            data["files"] = files_data
        if fds is not None:
            data["fds"] = fds
        return data

    @staticmethod
    def build_proc_info():
        pid = os.getpid()
        proc_info = psutil.Process(pid)
        return pid, proc_info

    def cpu_usage(self):
        cpu = None
        if self._config.include_cpu():
            cpu = {}
            usage = self.proc_info.cpu_percent()
            cpu["cpu_usage"] = usage
        return cpu

    def memory_usage(self):
        memory = None
        if self._config.include_memory():
            memory = {}
            usage = self.proc_info.memory_info().rss / 1000000
            memory["memory_usage"] = usage
        return memory

    def open_files(self):
        open_files = None
        if self._config.include_files():
            open_files = {}
            usage = len(self.proc_info.open_files())
            open_files["files"] = usage
        return open_files

    def file_descriptors(self):
        file_descriptors = None
        if self._config.include_fds():
            file_descriptors = {}
            if self._platform != "Windows":
                usage = self.proc_info.num_fds()
            else:
                usage = self.proc_info.num_handles()
            file_descriptors["fds"] = usage
        return file_descriptors

    def thread_count(self):
        thread_count = None
        if self._config.include_threads():
            thread_count = {}
            usage = len(self.proc_info.threads())
            thread_count["threads"] = usage
        return thread_count

    @staticmethod
    def __get_disk_usage():
        usage = shutil.disk_usage("..")
        usage = {"total": usage[0], "used": usage[1], "free": usage[2]}
        return usage

    def disk_usage(self):
        disk = None
        if self._config.include_disk():
            disk = {}
            disk_usage = HeartbeatMonitor.__get_disk_usage()
            total = disk_usage["total"] / 1000000000
            used = disk_usage["used"] / 1000000000
            free = disk_usage["free"] / 1000000000

            disk["disk_total"] = total
            disk["disk_used"] = used
            disk["disk_free"] = free
        return disk


if __name__ == "__main__":
    from ggcommons.config.manager.file_config_manager import FileConfigManager

    print(os.getcwd())
    config = FileConfigManager("PYTHON_TEST", "../../config_3.json")
    monitor = HeartbeatMonitor(config)
    print(monitor.pid)
    print(monitor.get_stats())

File: libs/java/src/main/java/com/mbreissi/edgecommons/heartbeat/HeartbeatMonitor.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.mbreissi.edgecommons.heartbeat;

import com.mbreissi.edgecommons.config.HeartbeatConfiguration;
import com.google.gson.JsonObject;
import oshi.PlatformEnum;
import oshi.SystemInfo;
import oshi.software.os.OSProcess;


/**
 * Monitors and tracks the heartbeat status of a Greengrass component.
 * Manages heartbeat statistics, health checks, and status reporting.
 */
public class HeartbeatMonitor
{
    HeartbeatConfiguration heartbeatConfiguration;
    SystemInfo si;
    OSProcess currentProc;
    OSProcess previousProc;
    long previousCpuTime = 0L;
    /**
     * Creates a new HeartbeatMonitor with the specified configuration.
     *
     * @param hbConfig The heartbeat configuration settings to use
     */
    public HeartbeatMonitor(HeartbeatConfiguration hbConfig)
    {
        heartbeatConfiguration = hbConfig;
        si = new SystemInfo();
        currentProc = new SystemInfo().getOperatingSystem().getCurrentProcess();
    }

    /**
     * Gets the current health statistics for this component.
     * Collects various metrics based on configuration settings.
     *
     * @return JsonObject containing health statistics
     */
    public JsonObject getStats()
    {
        JsonObject data = new JsonObject();
        updateMetrics();

        JsonObject cpuData = getCpuUsage();
        if (cpuData != null)
            data.add("cpu", cpuData);

        JsonObject memData = getMemoryUsage();
        if (memData != null)
            data.add("memory", memData);

        JsonObject diskData = getDiskUsage();
        if (diskData != null)
            data.add("disk", diskData);

        JsonObject threadData = getThreadCount();
        if (threadData != null)
            data.add("threads", threadData);

        JsonObject fileData = getFileCount();
        if (fileData != null)
            data.add("files", fileData);

        JsonObject fdData = getFdCount();
        if (fdData != null)
            data.add("fds", fdData);

        return data;
    }

    private JsonObject getCpuUsage()
    {
        JsonObject retVal = null;
        if (heartbeatConfiguration.includeCpu())
        {
            retVal = new JsonObject();
            double cpuUsage;
            cpuUsage = currentProc.getProcessCpuLoadBetweenTicks(previousProc)*100;
            if (SystemInfo.getCurrentPlatform() == PlatformEnum.WINDOWS)
                cpuUsage /= si.getHardware().getProcessor().getLogicalProcessorCount();
            retVal.addProperty("cpu_usage", cpuUsage);
        }
        return retVal;
    }

    private JsonObject getMemoryUsage()
    {
        JsonObject retVal = null;
        if (heartbeatConfiguration.includeMemory())
        {
            retVal = new JsonObject();
            retVal.addProperty("memory_usage", currentProc.getResidentSetSize()/1000000);
        }
        return retVal;
    }

    private JsonObject getDiskUsage()
    {
        JsonObject retVal = null;
        if (heartbeatConfiguration.includeDisk())
        {
            retVal = new JsonObject();
            // Mirror the Python/Rust libs: report the filesystem holding the parent of the working
            // dir, in Gigabytes (decimal, /1e9). used = total - free.
            java.io.File dir = new java.io.File("..");
            long total = dir.getTotalSpace();
            long free = dir.getFreeSpace();
            long used = total - free;
            retVal.addProperty("disk_total", total / 1.0e9);
            retVal.addProperty("disk_used", used / 1.0e9);
            retVal.addProperty("disk_free", free / 1.0e9);
        }
        return retVal;
    }

    private JsonObject getFdCount()
    {
        JsonObject retVal = null;
        if (heartbeatConfiguration.includeFds())
        {
            retVal = new JsonObject();
            // Open file descriptors via the Unix OS MXBean (Linux/macOS — the deploy target);
            // absent on Windows, where -1 signals "unavailable" rather than a bogus count.
            long fds = -1L;
            java.lang.management.OperatingSystemMXBean os =
                    java.lang.management.ManagementFactory.getOperatingSystemMXBean();
            if (os instanceof com.sun.management.UnixOperatingSystemMXBean unix)
            {
                fds = unix.getOpenFileDescriptorCount();
            }
            retVal.addProperty("fds", fds);
        }
        return retVal;
    }

    private JsonObject getThreadCount()
    {
        JsonObject retVal = null;
        if (heartbeatConfiguration.includeThreads())
        {
            retVal = new JsonObject();
            retVal.addProperty("threads", currentProc.getThreadCount());
        }
        return retVal;
    }

    private JsonObject getFileCount()
    {
        JsonObject retVal = null;
        if (heartbeatConfiguration.includeFiles())
        {
            retVal = new JsonObject();
            retVal.addProperty("files", currentProc.getOpenFiles());
        }
        return retVal;
    }

    public void updateMetrics()
    {
        previousProc = currentProc;
        currentProc = new SystemInfo().getOperatingSystem().getCurrentProcess();
    }
}
